### DISCUSSION TITLE
My FastAPI is not responding to request after waiting for a while, and requires Ctrl+C to restart

### DISCUSSION THREAD

**Discussion by laurence-lin** [fastapi/fastapi#8813] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, BackGroundTasks

app = FastAPI()


@app.post("/create", status_code=201, response_model=schemas.TaskResponseModel)
async def create_tasks(
    task: schemas.TrainingRequestModel, background_tasks: BackgroundTasks
):
    logging.info("{} Start executing background task at: ".format(datetime.now()))
    background_tasks.add_task(train_model, task)  # train model at background
    response = schemas.TaskResponseModel(task_id=task.task_id)
    logging.info("Created response")
    
    print("endpoint response:", response)
    return response
```

### Description

I'm building API application for machine learning models, and have endpoint "create" to input the training data then execute training.

After my Web APP send the request, my FastAPI waits for a long time and seems not receiving response. During when I send a simple request by typing `http://127.0.0.1:5000/docs` in my browser, and it waits a long time to run but not showing the swagger document. 

I assume FastAPI is working in background, maybe receiving the request but facing some problem, so I press Ctrl+C to stop the program. What strange happens is that as long as I press the terminate (Ctrl+C), the program starts running and receive my request to "create" endpoint. It's like FastAPI is stuck in some problem while receiving message, and I need to pause the work to make it continue.

Other from that problem mentioned above, when I sometimes typing `http://127.0.0.1:5000/docs`  in the browser after idle for a while, it waits for a long time with no response, and same I have to press Ctrl+C to then it continues running.

Please help me with this problem, any advice is appreciate. Thank  you!

### Operating System

Windows

### Operating System Details

Windows 10 Pro

### FastAPI Version

0.68.0

### Python Version

3.8.10

**Comment by anthonycepeda** [comment]
Hi, would you please provide more context regarding  schemas.TrainingRequestModel and schemas.TaskResponseModel?
Thanks

**Comment by laurence-lin** [comment]
> Hi, would you please provide more context regarding schemas.TrainingRequestModel and schemas.TaskResponseModel?
> Thanks

TrainingRequestModel:

```
class TrainingRequestModel(BaseModel):
    class TrainingData(BaseModel):
        text: str = Field(None, description="文本欄位")
        label: str = Field(None, description="類別欄位")

    task_id: str = Field(None, description="專案代號")
    ngram_size: int = Field(1, description="Ngram大小，建議值1~5")
    min_len: int = Field(20, description="文本長度下限")
    max_len: int = Field(2000, description="文本長度上限")
    test_size: float = Field(0.2, description="測試資料集比例，建議值0.2~0.5")
    efficiency: bool = Field(True, description="true為效率優先, false為正確率優先")
    dataset: Optional[List[TrainingData]] = Field(
        None, description="資料集列表，建議至少200筆資料")
```

TaskResponseModel:

```
class TaskResponseModel(BaseModel):
    status: bool = Field(True, description="回傳狀態")
    task_id: str = Field(None, description="專案代號")
```

As above, thank you

**Comment by anthonycepeda** [comment]
Hi @laurence-lin,  the problem seems to be when the function you pass to the [backgrond_task](https://fastapi.tiangolo.com/tutorial/background-tasks/):  `train_model`, is async.

If you **take `async` keyword out** of your **training_model function**, you should be good to go.

Thanks

**Comment by laurence-lin** [comment]
> Hi @laurence-lin, the problem seems to be when the function you pass to the [backgrond_task](https://fastapi.tiangolo.com/tutorial/background-tasks/): `train_model`, is async.
> 
> If you **take `async` keyword out** of your **training_model function**, you should be good to go.
> 
> Thanks

Thank you for response, I've remove the `async` from endpoint function `create`, but the problem remains.
My `train_model` function is without `async` quote:

```
def train_model(train_setting):
      train_data = train_setting.train_data
      train_param = train_setting.train_param

      y = train_data["label"]
      x = train_data.drop("label", axis = 1)
      model = Xgboost(**train_param)
      model.fit(x, y)

      model.save("model.pkl")

       return None
```

I'm wondering should it be something do to the background tasks? Should I use `await` before train_model function?

**Comment by Kludex** [comment]
No.

I guess you're running the server with a single worker. Try to add more workers.

**Comment by upeshphable** [comment]
@Kludex I have multiple workers in my application but still the background task blocks the entire application. All requests keep waiting until the task finishes.

**Comment by anthonycepeda** [comment]
> @Kludex I have multiple workers in my application but still the background task blocks the entire application. All requests keep waiting until the task finishes.

Hi,

It's the function in background `async`? 
If so, try taking it out.

Thanks

**Comment by upeshphable** [comment]
@anthonycepeda It has async. The documentation says FastAPI will know how to handle it correctly. 

[https://fastapi.tiangolo.com/tutorial/background-tasks/](url)

```Create a task function¶
Create a function to be run as the background task.

It is just a standard function that can receive parameters.

It can be an async def or normal def function, FastAPI will know how to handle it correctly.

In this case, the task function will write to a file (simulating sending an email).
```

Is this a defect or just a documentation cleanup?

**Comment by upeshphable** [comment]
I tried with removing async and wrapping up the task in a regular function. It started to lock the DB.

**Comment by john-jaraceski** [comment]
Hi. I really don't know if you are still dealing with this problem... I'm facing the same problem in my application and I'm using _subprocess_ rather than _threads_. And the FastAPI seems to no longer freezing.

https://docs.python.org/3/library/multiprocessing.html

**Comment by nickwong64** [comment]
I am also facing this problem. The fastapi sometimes does not respond and if I try to hit 'enter' or 'ctrl-c' in the console. The API resumes. 
Any idea how to fix this? 
Thanks.

**Reply by john-jaraceski** [reply]
Maybe, your freezing problem is related with something else. I faced this situation with a regular expression used to validate input data submitted to API. Are you using regular expressions?

An important note, the same regular expression worked perfectly in both Java and .NET but presented an infinite recursion in Python (and the freezing problem).

**Reply by nickwong64** [reply]
Looked around the internet and found the workaround:
I am calling an API which takes some times to respond, I changed the **'async def' to 'def'** and the problem is gone.
I am not sure why calling this long responding API need to use **non-async def**. But seems it works. 
As I know async should be used to call long running process (in javascript case) so that it won't block other process, not sure why it shoudn't be used in Python FastAPI.

**Reply by john-jaraceski** [reply]
You need to use ```non-async``` to force FastAPI to use a thread to process this request. The ```async``` functions are processed in the main thread and they should be used to process fast request.

**Reply by zhaoyangkun** [reply]
I am facing the same problem, sometimes fastapi server is not responding in Win10. Now I change uvicorn to hypercorn as the asgi server to temporarily solve this problem.

**Comment by mht2953658596** [comment]
I am also facing this problem. The fastapi sometimes does not respond and if I try to hit 'enter' or 'ctrl-c' in the console. The API resumes.
Any idea how to fix this?
Thanks.

**Reply by john-jaraceski** [reply]
Maybe, your freezing problem is related with something else. I faced this situation with a regular expression used to validate input data submitted to API. Are you using regular expressions?

An important note, the same regular expression worked perfectly in both Java and .NET but presented an infinite recursion in Python (and the freezing problem).

**Comment by anthonycepeda** [comment, marked as answer]
Hello,

I have tested multiple scenarios and the situation described on this issue only arises when:
a) using async over a function that should not be
b) running the server on a single worker

In this particular case, I think `train_model` function used in a background task, should not be async

**Reply by daiz93** [reply]
On windows going to task manager and stoped fastapi.exe worked for me.

**Comment by zaynekomichi** [comment]
removing async helped me

**Reply by rodrigos0ares** [reply]
@app.get("/", tags=["root"])
def read_root():
    return {'status': 'OK'}
    
Removing the async from this function did the trick for me.

**Comment by estebanx64** [comment]
The background task in fastapi are made to small tasks whose not gonna take minutes and are heavy in cpu/memory. I'm guessing if you are training a model this is the case of a heavy task. you should use another tools design for it, like airflow, celery among others.

Maybe you can give us more detail about what are you training?

**Comment by bogdan-coman-uv** [comment]
This is a python problem, not related to FastApi. Your training task is most probably heavily CPU bound so you background task will block your current worker process. Try running the training task in a separate subprocess. Take a look at https://docs.python.org/3/library/asyncio-subprocess.html
